Create a deploy script in foundry script.

cuddle-dev opened the pull request:
Resolves #

<!--
- You must link the issue number e.g. "Resolves #1234"
- Please do not replace the keyword "Resolves" https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

cuddle-dev committed:
Create slither.yml

cuddle-dev committed:
Create slither.yml

cuddle-dev committed:
Create slither.yml

cuddle-dev committed:
move abdk to lib

cuddle-dev committed:
move abdk to lib

cuddle-dev committed:
move abdk to lib

cuddle-dev committed:
move abdk to lib

cuddle-dev committed:
upgrade solidity version

cuddle-dev committed:
upgrade solidity version

cuddle-dev committed:
upgrade solidity version

cuddle-dev committed:
upgrade solidity version

cuddle-dev committed:
update slither config

cuddle-dev committed:
update slither config

cuddle-dev committed:
update slither config

cuddle-dev committed:
update slither config

cuddle-dev committed:
add detectors to exclude list

cuddle-dev committed:
add detectors to exclude list

cuddle-dev committed:
add detectors to exclude list

cuddle-dev committed:
add detectors to exclude list

cuddle-dev committed:
Fix filter_path

cuddle-dev committed:
Fix filter_path

cuddle-dev committed:
Fix filter_path

cuddle-dev committed:
Fix filter_path

cuddle-dev committed:
wrong path

cuddle-dev committed:
wrong path

cuddle-dev committed:
wrong path

cuddle-dev committed:
wrong path

cuddle-dev committed:
finalize

cuddle-dev committed:
finalize

cuddle-dev committed:
finalize

cuddle-dev committed:
finalize

cuddle-dev committed:
delete out directory

cuddle-dev committed:
delete out directory

cuddle-dev committed:
Merge branch 'development' into integrate-slither

cuddle-dev committed:
Merge branch 'development' into integrate-slither

cuddle-dev committed:
Remove unnecessary files

cuddle-dev committed:
Remove unnecessary files

cuddle-dev committed:
Remove unnecessary files

cuddle-dev committed:
Remove unnecessary files

cuddle-dev committed:
remove dependencies for non-existing contracts

cuddle-dev committed:
remove dependencies for non-existing contracts

cuddle-dev committed:
remove dependencies for non-existing contracts

cuddle-dev committed:
remove dependencies for non-existing contracts

cuddle-dev committed:
Revert due to dapp changes

cuddle-dev committed:
Revert due to dapp changes

cuddle-dev committed:
Revert due to dapp changes

cuddle-dev committed:
Revert due to dapp changes

cuddle-dev committed:
update slither config

cuddle-dev committed:
update slither config

cuddle-dev committed:
update slither config

cuddle-dev committed:
update slither config

cuddle-dev committed:
fix unchecked-transfer issue

cuddle-dev committed:
fix unchecked-transfer issue

cuddle-dev committed:
fix unchecked-transfer issue

cuddle-dev committed:
fix unchecked-transfer issue

cuddle-dev committed:
performs a multiplication on the result of division

cuddle-dev committed:
performs a multiplication on the result of division

cuddle-dev committed:
performs a multiplication on the result of division

cuddle-dev committed:
fix too many digits and constant

cuddle-dev committed:
Update slither.config.json

cuddle-dev committed:
Update slither.config.json

cuddle-dev committed:
Update slither.config.json

cuddle-dev committed:
Update slither.config.json

cuddle-dev committed:
missing inheritance

cuddle-dev committed:
missing inheritance

cuddle-dev committed:
missing inheritance

cuddle-dev committed:
missing inheritance

hashedMae committed:
slither issues

hashedMae committed:
slither issues

hashedMae committed:
slither issues

hashedMae committed:
slither issues

hashedMae committed:
up to slither 204

hashedMae committed:
up to slither 204

hashedMae committed:
up to slither 204

hashedMae committed:
recheck slither

hashedMae committed:
recheck slither

hashedMae committed:
recheck slither

cuddle-dev committed:
Merge branch 'development' into integrate-slither

hashedMae committed:
Merge branch 'Issue-284' of into integrate-slither

hashedMae committed:
Cleanup Slither Flags

hashedMae committed:
Cleanup Slither Flags

hashedMae committed:
Cleanup Slither Flags

hashedMae committed:
complilation errors

hashedMae committed:
complilation errors

hashedMae committed:
complilation errors

hashedMae committed:
rentrancy, unitialized variables, return values

hashedMae committed:
rentrancy, unitialized variables, return values

hashedMae committed:
rentrancy, unitialized variables, return values

hashedMae committed:
rentrancy in addLiquidity()

hashedMae committed:
gas savings in MCV2, reentrancy in CollectableDust

hashedMae committed:
gas savings in MCV2, reentrancy in CollectableDust

hashedMae committed:
gas savings in MCV2, reentrancy in CollectableDust

hashedMae committed:
gas optimization MCV2

hashedMae committed:
_updatePool outside loop

hashedMae committed:
Revert "_updatePool outside loop"

This reverts commit bc2a9e5f27714b1ddc3c0c5fddb0fe7bbeeae29b.

hashedMae committed:
fix tests

hashedMae committed:
fix tests

hashedMae committed:
fix tests

hashedMae committed:
fix tests

hashedMae committed:
Change Contract names to reflect Refactor

hashedMae committed:
Change Contract names to reflect Refactor

hashedMae committed:
Change Contract names to reflect Refactor

hashedMae committed:
Change Contract names to reflect Refactor

hashedMae committed:
Revert "Change Contract names to reflect Refactor"

This reverts commit a5e8435e1d52b05e1865434a3af10eda796cd38b.

hashedMae committed:
Revert "Change Contract names to reflect Refactor"

This reverts commit a5e8435e1d52b05e1865434a3af10eda796cd38b.

hashedMae committed:
Revert "Revert "Change Contract names to reflect Refactor""

This reverts commit 5d4813f9bca77aa2e9c8a078eb63dc7b63840c2c.

hashedMae committed:
Revert "Revert "Change Contract names to reflect Refactor""

This reverts commit 5d4813f9bca77aa2e9c8a078eb63dc7b63840c2c.

hashedMae committed:
Renamings

hashedMae committed:
Renamings

hashedMae committed:
Renamings

hashedMae committed:
Renamings

hashedMae committed:
renaming

hashedMae committed:
renaming

hashedMae committed:
renaming

hashedMae committed:
renaming

hashedMae committed:
Merge branch 'integrate-slither' of https://github.com/ubiquity/ubiquity-dollar into integrate-slither

hashedMae committed:
rename refactor in progress

hashedMae committed:
rename refactor in progress

hashedMae committed:
finish name refactor

hashedMae committed:
finish name refactor

hashedMae committed:
json error

hashedMae committed:
gas optimization contstructor

hashedMae committed:
gas optimization contstructor

hashedMae committed:
bug fix

hashedMae committed:
bug fix

hashedMae committed:
Slither ignore payable.call in CollectableDust

hashedMae committed:
Slither ignore payable.call in CollectableDust

hashedMae committed:
slither-disable-next-line low-level-calls

hashedMae committed:
slither-disable-next-line low-level-calls

hashedMae committed:
solidity version, remove empty file

hashedMae committed:
solidity version, remove empty file

hashedMae committed:
fixed bug deploy scripts

hashedMae committed:
fixed bug deploy scripts

hashedMae committed:
same

hashedMae committed:
same

hashedMae committed:
again

hashedMae committed:
again

hashedMae committed:
build-test flow bugs

hashedMae committed:
build-test flow bugs

hashedMae committed:
build-test scripts bug

hashedMae committed:
build-test scripts bug

hashedMae committed:
testFixes

hashedMae committed:
testFixes

cuddle-dev committed:
Merge branch 'feat/406-protocol-deploy-script' into integrate-slither

cuddle-dev committed:
fix: resolve conflicts

cuddle-dev committed:
fix: resolve conflicts

cuddle-dev committed:
fix: slither issues

cuddle-dev committed:
fix: slither issues

cuddle-dev committed:
fix: remove unused return

cuddle-dev committed:
fix: remove unused return

cuddle-dev committed:
fix: loop

cuddle-dev committed:
fix: loop

cuddle-dev committed:
fix: uninitialized local

cuddle-dev committed:
fix: uninitialized local

cuddle-dev committed:
fix: MUST revert

cuddle-dev committed:
fix: MUST revert

cuddle-dev committed:
Merge pull request #302 from ubiquity/integrate-slither

Fix the slither issues

0x4007 committed:
fix: pr review easy adjustments

0x4007 committed:
fix: pr review easy adjustments

0x4007 committed:
fix: "smart contract" is two words!

0x4007 committed:
fix: "smart contract" is two words!

hashedMae committed:
fix(contracts): spelling

hashedMae committed:
fix(contracts): spelling

hashedMae committed:
fix(dapp): deploy scripts contract name

hashedMae committed:
fix(dapp): deploy scripts contract name

hashedMae committed:
fix(contracts): error capitalization

hashedMae committed:
fix(contracts): error capitalization

0x4007 committed:
Merge pull request #424 from ubiquity/pr-302-follow-up

fix(contracts): spelling

hashedMae committed:
feat: solidity deployment script

hashedMae committed:
feat: solidity deployment script

hashedMae committed:
feat: solidity scripting devlopment deploy

hashedMae committed:
feat: solidity scripting devlopment deploy

hashedMae committed:
feat: devlopment deploy twap fix

hashedMae committed:
feat: devlopment deploy twap fix

hashedMae committed:
feat: deploy script
added bash scripts for ease of use

hashedMae committed:
feat: deploy script
added bash scripts for ease of use

0x4007 commented on package.json:
```diff
@@ -36,7 +36,7 @@
   },
   "lint-staged": {
-    "*.{js,jsx,ts,tsx,json,sol}": [
+    "*.{js,jsx,ts,tsx,json}": [
```
1. Couldn't this PR be rebased on a more current commit from `development` so that I don't need to see redundant changes from all the other PRs I reviewed? I'm trying now locally.

1. I very quickly rebased to development and opened up this draft PR with a few extra commit changes (we should delete my draft PR when @hashedMae is done here) https://github.com/ubiquity/ubiquity-dollar/pull/471

1. It looks like there were a handful of deleted/renamed/moved files and I'm becoming disoriented on how to proceed. It looks mostly correct but I didn't anticipate the changes related to the UbiquiStick files.

0x4007 commented on packages/contracts/package.json:
```diff
@@ -23,8 +23,8 @@
     "build": "forge build",
-    "test": "ts-node ./scripts/runner/runner.ts",
-    "slither:run": "slither . --compile-force-framework foundry",
+    "test": "forge test --fork-url https://cloudflare-eth.com",
```
```suggestion
    "test": "ts-node ./scripts/runner/runner.ts",
```

I know that there's something wrong with what commit this is based on because the `runner.ts` script is definitely the correct answer here. @hashedMae perhaps you can spend some time looking into rebasing off of the correct commit?

0x4007 commented on packages/contracts/scripts/bash/1_StartAnvil.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+# Starts Anvil with mainnet fork explicitly using default accounts
+anvil -f https://cloudflare-eth.com -m 'test test test test test test test test test test test junk'
```
This should use our `runner.ts` I referenced above. You can use `npx tsx` or we can just add `tsx` as a dependency to automatically consume and run TypeScript files.

0x4007 commented on packages/contracts/scripts/deploy/dollar/solidityScripting/TWAPOracle.s.sol:
```diff
@@ -0,0 +1,26 @@
+contract TWAPScript is GovernanceScript {
+	function run() public virtual override {
+		super.run();
+		uint256 deployerPrivateKey = vm.envUint("PRIVATE_KEY");
```
I wish that we could dynamically parse the code and render a list of required environment variables in the `.env.example` file!

0x4007 commented on packages/contracts/slither.config.json:
```diff
@@ -1,9 +1,10 @@
   "exclude_high": false,
   "disable_color": false,
-  "filter_paths": "(lib/|test/)",
+  "filter_paths": "lib|test|src/dollar/mocks|src/ubiquistick",
```
Why is slither skipping ubiquistick? I also remember seeing this on another PR some time ago.

cuddle-dev replied:
Ideally it should be included in slither checks but I figured out it requires a lot of changes. Fixing slither issue would take lots of time rather than re-implementing them properly. so I made it excluded. In the near future, we should include again.

0x4007 commented on packages/contracts/src/dollar/DirectGovernanceFarmer.sol:
```diff
@@ -428,9 +368,9 @@ contract DirectGovernanceFarmer is ReentrancyGuard {
+                || token == ubiquityDollar
+        );
     }
 }
```
Looks like formatting got messed up on this file. Which one is the correct one?

hashedMae replied:
I changed it so each OR statement is on it's own line which meets the style guide.

0x4007 commented on packages/contracts/src/dollar/ERC1155Ubiquity.sol:
```diff
@@ -24,15 +24,15 @@ contract ERC1155Ubiquity is ERC1155, ERC1155Burnable, ERC1155Pausable {
     modifier onlyMinter() {
         require(
-            manager.hasRole(manager.GOVERNANCE_TOKEN_MINTER_ROLE(), msg.sender),
+            manager.hasRole(manager.UBQ_MINTER_ROLE(), msg.sender),
```
Looks like this is a partial rename from our global name changes PR. `GOVERNANCE_TOKEN_MINTER_ROLE` appears to be updated according to the new naming convention. Which is correct?

hashedMae replied:
Changed them all to `GOVERNANCE_TOKEN_MINTER_ROLE`

0x4007 commented on packages/contracts/src/dollar/ERC20Ubiquity.sol:
```diff
@@ -34,15 +29,15 @@ contract ERC20Ubiquity is
     modifier onlyMinter() {
         require(
-            manager.hasRole(manager.GOVERNANCE_TOKEN_MINTER_ROLE(), msg.sender),
+            manager.hasRole(manager.UBQ_MINTER_ROLE(), msg.sender),
```
Same as above.

hashedMae replied:
Changed them all to GOVERNANCE_TOKEN_MINTER_ROLE

0x4007 commented on packages/contracts/src/dollar/StakingFormulas.sol:
```diff
@@ -1,8 +1,8 @@
 
 import "./StakingShare.sol";
-import "./libs/ABDKMathQuad.sol";
+import "abdk-libraries-solidity/ABDKMathQuad.sol";
```
I thought that the updated Solidity version made ABDK library obsolete @0xcodercrane please advise.

0x4007 commented on packages/contracts/src/dollar/UbiquityChef.sol:
```diff
@@ -253,47 +286,53 @@ contract UbiquityChef is ReentrancyGuard {
-            governanceToken.safeTransfer(_to, governanceBalance);
+        uint256 governanceBal = governance.balanceOf(address(this));
+        if (_amount > governanceBal) {
+            governance.safeTransfer(_to, governanceBal);
```
```suggestion
            governance.safeTransfer(_to, governanceBalance);
```

We should strive to never go shorthand in order to maximize codebase readability.

hashedMae replied:
Changed to governanceBalance

0x4007 commented on packages/contracts/src/dollar/UbiquityFormulas.sol:
```diff
@@ -1,119 +1,123 @@
+    /// @notice uLP = uBOND * currentShareValue / targetPrice
+    // uLP = A * V / T
+    function redeemShares(
+        uint256 uBOND,
```
This doesn't seem right to refer to it as `uBOND` for several reasons.
1. The original "uBOND" technically was a non-expiring Ubiquity Credit NFT
2. This is referring to "bonding shares" which is sort of a hidden token for internal accounting. 
3. We renamed "bonding" to "staking"

This should be named with its full name, like stakingShare or something.

0x4007 commented on packages/contracts/src/dollar/core/CreditNFT.sol:
```diff
@@ -1,14 +1,15 @@
+import "../interfaces/ICreditNFT.sol";
 
-/// @title A Credit NFT redeemable for dollars with an expiry block number
+/// @title A creditNFT redeemable for dollars with an expiry block number
```
```suggestion
/// @title A Credit NFT redeemable for Dollars with an expiry block number
```

Perhaps it should be reverted to the old comment style.

hashedMae replied:
Changed to `CreditNFT` as that is the contract name. `creditNFT` was inappropriate as that would be a specific instance referenced as a variable.

0x4007 commented on packages/contracts/src/dollar/core/CreditNFT.sol:
```diff
@@ -55,30 +56,30 @@ contract CreditNFT is ERC1155Ubiquity {
         //insert new relevant block number if it doesn't exist in our list
         // (linked list implementation won't insert if dupe)
-        _sortedBlockNumbers.pushBack(expiryBlockNumber);
+        require(_sortedBlockNumbers.pushBack(expiryBlockNumber));
```
Was this (lack of `require`) a vulnerability picked up by Slither etc?

hashedMae replied:
Yes, both the EnumerableSet and StructurededLinkList will return a bool for functions that alter the sets. Slither was calling these out as being ignored and a require was the simplest way to ensure a revert on failure.

0x4007 commented on packages/contracts/src/dollar/core/CreditNFT.sol:
```diff
@@ -55,30 +56,30 @@ contract CreditNFT is ERC1155Ubiquity {
     }
 
-    /// @notice Burn an amount of Credit NFT expiring at a certain block from
+    /// @notice Burn an amount of creditNFTs expiring at a certain block from
```
Same as earlier comment.

hashedMae replied:
Changed to `CreditNFT` as that is the contract name

0x4007 commented on packages/contracts/src/dollar/core/CreditNFT.sol:
```diff
@@ -91,22 +92,24 @@ contract CreditNFT is ERC1155Ubiquity {
             if (currentBlockNumber > block.number) {
                 //put the key back in since we popped, and end loop
-                _sortedBlockNumbers.pushFront(currentBlockNumber);
+                require(_sortedBlockNumbers.pushFront(currentBlockNumber));
```
Also same as earlier `require` comment.

hashedMae replied:
Both the EnumerableSet and StructurededLinkList will return a bool for functions that alter the sets. Slither was calling these out as being ignored and a require was the simplest way to ensure a revert on failure.

0x4007 commented on packages/contracts/src/dollar/core/CreditNFTManager.sol:
```diff
@@ -389,9 +398,12 @@ contract CreditNFTManager is ERC165, IERC1155Receiver {
+                    manager.getExcessDollarsDistributor(address(this)),
+                    excessDollars
+                ),
+                "UbiquityDollar: Transfer failed"
```
```suggestion
                "Dollar: Transfer failed"
```

I feel that writing "Ubiquity" is redundant within the Ubiquity Dollar codebase so I err on the side of removing when I see "Ubiquity"

0x4007 commented on packages/contracts/src/dollar/core/CreditNFTRedemptionCalculator.sol:
```diff
@@ -1,11 +1,11 @@
-import "../libs/ABDKMathQuad.sol";
-import "./CreditNFT.sol";
 import "./UbiquityDollarManager.sol";
+import "abdk-libraries-solidity/ABDKMathQuad.sol";
```
Refer to abdk comment above.

0x4007 commented on packages/contracts/src/dollar/core/CreditRedemptionCalculator.sol:
```diff
@@ -18,7 +18,7 @@ contract CreditRedemptionCalculator is ICreditRedemptionCalculator {
         require(
             manager.hasRole(manager.INCENTIVE_MANAGER_ROLE(), msg.sender),
-            "CreditCalc: not admin"
+            "UARCalc: not admin"
```
```suggestion
            "CreditCalculator: not admin"
```

Looks like the old naming convention was correct.

hashedMae replied:
done

0x4007 commented on packages/contracts/src/dollar/core/CreditRedemptionCalculator.sol:
```diff
@@ -28,14 +28,14 @@ contract CreditRedemptionCalculator is ICreditRedemptionCalculator {
-    /// @notice set the constant for Credit Token calculation
-    /// @param coef new constant for Credit Token calculation in ETH format
+    /// @notice set the constant for uAR calculation
+    /// @param coef new constant for uAR calculation in ETH format
```
Also old naming convention being correct
```suggestion
    /// @notice set the constant for Credit token calculation
    /// @param coef new constant for Credit token calculation in ETH format
```

hashedMae replied:
Done

0x4007 commented on packages/contracts/src/dollar/core/DollarMintCalculator.sol:
```diff
@@ -20,17 +20,16 @@ contract DollarMintCalculator is IDollarMintCalculator {
     }
 
-    /// @notice returns (TWAP_PRICE  -1) * Ubiquity_Dollar_Total_Supply
+    /// @notice returns (TWAP_PRICE  -1) * UAD_Total_Supply
```
```suggestion
    /// @notice returns (TWAP_PRICE  -1) * Dollar_Total_Supply
```

We should never refer to the tokens by their symbols in the codebase because they are not as clear for new developers as writing their utility (Dollar, Credit etc) also we can (and already have) updated the symbols.

hashedMae replied:
done

0x4007 commented on packages/contracts/src/dollar/core/DollarMintExcess.sol:
```diff
@@ -1,16 +1,16 @@
-import "../libs/ABDKMathQuad.sol";
 import "./UbiquityDollarManager.sol";
+//import "./SushiSwapPool.sol";
+import "abdk-libraries-solidity/ABDKMathQuad.sol";
```
abdk comment above

0x4007 commented on packages/contracts/src/dollar/core/DollarMintExcess.sol:
```diff
@@ -37,25 +37,25 @@ contract DollarMintExcess is IDollarMintExcess {
             address treasuryAddress = manager.treasuryAddress();
 
-            // curve UbiquityDollar-3CRV liquidity pool
+            // curve uAD-3CRV liquidity pool
```
```suggestion
            // curve UbiquityDollar-3CRV liquidity pool
```

Looks like the old changes were correct from the renames. I included "Ubiquity" here to reduce ambiguity because its referring to other protocol tokens in the same sentence.

hashedMae replied:
done

0x4007 commented on packages/contracts/src/dollar/core/DollarMintExcess.sol:
```diff
@@ -37,25 +37,25 @@ contract DollarMintExcess is IDollarMintExcess {
                 treasuryAddress, fiftyPercent
             );
-            // convert Ubiquity Dollar to GovernanceToken-DollarToken LP on sushi and burn them
+            // convert uAD to uGOV-UAD LP on sushi and burn them
```
Old changes comment

hashedMae replied:
done

0x4007 commented on packages/contracts/src/dollar/core/TWAPOracleDollar3pool.sol:
```diff
@@ -1,31 +1,28 @@
     address public immutable token1;
-    uint256 public price0Average;
-    uint256 public price1Average;
+    uint256 public uADPrice;
```
Old changes comment again. Perhaps this should be UbiquityDollarPrice given that its dealing with other protocol tokens as well in this code.

hashedMae replied:
done

0x4007 commented on packages/contracts/src/dollar/core/TWAPOracleDollar3pool.sol:
```diff
@@ -35,11 +32,11 @@ contract TWAPOracleDollar3pool {
-        priceCumulativeLast = IMetaPool(_pool).get_price_cumulative_last();
-        pricesBlockTimestampLast = IMetaPool(_pool).block_timestamp_last();
+        ///priceCumulativeLast = IMetaPool(_pool).get_price_cumulative_last();
+        ///pricesBlockTimestampLast = IMetaPool(_pool).block_timestamp_last();
```
```suggestion

```

Looks like these should be removed?

0x4007 commented on packages/contracts/src/dollar/core/UbiquityDollarManager.sol:
```diff
@@ -21,10 +21,8 @@ import "./TWAPOracleDollar3pool.sol";
-    bytes32 public constant GOVERNANCE_TOKEN_BURNER_ROLE =
-        keccak256("GOVERNANCE_TOKEN_BURNER_ROLE");
+    bytes32 public constant UBQ_MINTER_ROLE = keccak256("UBQ_MINTER_ROLE");
+    bytes32 public constant UBQ_BURNER_ROLE = keccak256("UBQ_BURNER_ROLE");
```
Looks like the role names were reverted to pre-updated-renames.

hashedMae replied:
fixed

0x4007 commented on packages/contracts/src/dollar/core/UbiquityDollarManager.sol:
```diff
@@ -81,7 +79,7 @@ contract UbiquityDollarManager is AccessControl {
 
         TWAPOracleDollar3pool oracle = TWAPOracleDollar3pool(twapOracleAddress);
-        oracle.update();
+        ///oracle.update();
```
```suggestion

```

We shouldn't update the oracle here?

hashedMae replied:
On local testnet, it fails unless a trade has taken place, like the `metapool.get_price_cumulative_last()` in the constructor. I don't understand why the behaviour works on mainnet but not a local testnet.

hashedMae replied:
The values are also null until a trade takes place so we're not losing anything by not calling it here.

0x4007 replied:
@zgorizzo69 do you have any remarks on this?

0x4007 commented on packages/contracts/src/dollar/interfaces/IStakingShare.sol:
```diff
@@ -1,5 +1,6 @@
 // SPDX-License-Identifier: Apache-2.0
-pragma solidity ^0.8.3;
+pragma solidity ^0.8.16;
+pragma solidity ^0.8.16;
```
I fixed this error already in another PR.
```suggestion
pragma solidity ^0.8.16;
```

hashedMae replied:
fixed again

0x4007 commented on packages/contracts/src/dollar/mocks/MockERC4626.sol:
```diff
@@ -43,7 +43,7 @@ contract MockERC4626 is ERC4626 {
         require(shares <= maxMint(receiver), "ERC4626: mint more than max");
 
-        uint256 assets = previewMint(shares);
+        assets = previewMint(shares);
```
Seems suspicious that uint256 was removed here. Why?

hashedMae replied:
`uint256 assets` is declared in the return statement instead of just `return(uint256)`

0x4007 commented on packages/contracts/src/dollar/mocks/MockERC4626.sol:
```diff
@@ -57,7 +57,7 @@ contract MockERC4626 is ERC4626 {
         require(assets <= maxWithdraw(owner), "ERC4626: withdraw more than max");
 
-        uint256 shares = previewWithdraw(assets);
+        shares = previewWithdraw(assets);
```
Seems suspicious that uint256 was removed here. Why?

hashedMae replied:
declared in return

0x4007 commented on packages/contracts/src/dollar/transfer-hooks/CurveDollarIncentive.sol:
```diff
@@ -1,12 +1,12 @@
 import "../interfaces/IUbiquityGovernance.sol";
 import "../interfaces/IIncentive.sol";
-import "../libs/ABDKMathQuad.sol";
+import "abdk/ABDKMathQuad.sol";
```
Refer to abdk comments.

0x4007 commented on packages/contracts/src/dollar/utils/CollectableDust.sol:
```diff
@@ -24,7 +24,7 @@ abstract contract CollectableDust is ICollectableDust {
             "collectable-dust::token-is-part-of-the-protocol"
         );
-        _protocolTokens.add(_token);
+        require(_protocolTokens.add(_token));
```
Curious about this added `require`

hashedMae replied:
Both the EnumerableSet and StructurededLinkList will return a bool for functions that alter the sets. Slither was calling these out as being ignored and a require was the simplest way to ensure a revert on failure.

0x4007 commented on packages/contracts/src/dollar/utils/CollectableDust.sol:
```diff
@@ -33,24 +33,25 @@ abstract contract CollectableDust is ICollectableDust {
             "collectable-dust::token-not-part-of-the-protocol"
         );
-        _protocolTokens.remove(_token);
+        require(_protocolTokens.remove(_token));
```
Curious about this added `require`

hashedMae replied:
both the EnumerableSet and StructurededLinkList will return a bool for functions that alter the sets. Slither was calling these out as being ignored and a require was the simplest way to ensure a revert on failure.

0x4007 commented on packages/contracts/test/dollar/ERC20Ubiquity.t.sol:
```diff
@@ -204,7 +204,7 @@ contract ERC20UbiquityTest is LocalTestHelper {
         vm.prank(admin);
         UbiquityDollarManager(dollar_manager_addr).grantRole(
-            keccak256("GOVERNANCE_TOKEN_BURNER_ROLE"), burner
+            keccak256("UBQ_BURNER_ROLE"), burner
```
Refer to changed role names comments above.

hashedMae replied:
Fixed

0x4007 commented on packages/contracts/test/dollar/ERC20Ubiquity.t.sol:
```diff
@@ -28,8 +27,9 @@ contract ERC20UbiquityTest is LocalTestHelper {
-    function setUp() public {
-        dollar_manager_addr = helpers_deployUbiquityDollarManager();
+    function setUp() public override {
+        super.setUp();
```
Weird that GitHub renders this super in italics. I presume that its normal unicode symbols if it passed the linters but it stood out to me enough to point attention to.

hashedMae replied:
Presumption is correct

0x4007 commented on packages/contracts/test/dollar/StakingFormulas.t.sol:
```diff
@@ -3,7 +3,7 @@ pragma solidity ^0.8.3;
 import "../../src/dollar/StakingFormulas.sol";
 import "../../src/dollar/StakingShare.sol";
-import "../../src/dollar/libs/ABDKMathQuad.sol";
+import "abdk/ABDKMathQuad.sol";
```
Refer to abdk comments above.

0x4007 commented on packages/contracts/test/dollar/core/CreditNFTManager.t.sol:
```diff
@@ -209,7 +205,7 @@ contract CreditNFTManagerTest is LocalTestHelper {
         );
         UbiquityDollarManager(dollarManagerAddress).grantRole(
-            keccak256("GOVERNANCE_TOKEN_MINTER_ROLE"), creditNFTManagerAddress
+            keccak256("UBQ_MINTER_ROLE"), creditNFTManagerAddress
```
Refer to updated role names comments above.

hashedMae replied:
Fixed

0x4007 commented on packages/contracts/test/dollar/core/DollarMintExcess.t.sol:
```diff
@@ -77,8 +75,9 @@ contract DollarMintExcessTest is LocalTestHelper {
+        );
         vm.mockCall(
             _metaPoolAddress,
             abi.encodeWithSelector(IMetaPool.exchange.selector),
```
I just realized that GitHub UI italicizes a handful of words. I count `super` and `selector` already.

0x4007 commented on packages/contracts/test/helpers/LiveTestHelper.sol:
```diff
@@ -107,8 +107,8 @@ contract LiveTestHelper is Test {
         manager.setStakingContractAddress(address(stakingV1));
-        manager.grantRole(manager.GOVERNANCE_TOKEN_MINTER_ROLE(), address(stakingV1));
-        manager.grantRole(manager.GOVERNANCE_TOKEN_MINTER_ROLE(), address(stakingShareV1));
+        manager.grantRole(manager.UBQ_MINTER_ROLE(), address(stakingV1));
```
Refer to updated roles comments above.

hashedMae replied:
fixed

0x4007 reviewed (changes requested):
After doing a draft PR to try and rebase from `development` and deeply going through this current pull request, my conclusion is that most of the renames (from the grand codebase rename PR) have been included but not all of them. Especially the staking related contracts seem to be out-of-date including the `UBQ_MINTER` role scattered about (that should be `GOVERNANCE_TOKEN_MINTER_ROLE`.)

The code mostly looks up-to-date based on the other PRs I reviewed in the last month or so. But it was unusual to me that some files like the `package.json` was out of date with the RPC url (should be selecting from the RPC selector script.)

I just wish that we didn't include redundant changes (I know that the Solidity pragma upgrades were from Slither, and not directly related to writing the deploy script) we should have possibly rebased from a newer `development` commit and tried to limit the amount of changes in a single PR. I'm actually not 100% sure what the best solution is but we should strive to come up with a strategy as part of our Git flow as it is very tedious for me to see the same changes from multiple PRs over and over again.

Another bit I noticed is the continued inclusion of ABDK libraries, when from what I understand, the latest version of Solidity we just upgraded to makes the use of these obsolete.

Good job team!

cuddle-dev reviewed (commented):
I have a couple of questions for @hashedMae . 

1/ I am just wondering why DevelopmentDeploy.s.sol is inherited from StakingScript, Stakingscript from FormulaScript, etc

DevelopmentDeploy.s.sol <= StakingScript <= FormulaScript <= CreditScript <= TWAPScript ... 
What I am seeing the reason is that you want to run only `super.run()` ??? I want to clarify before going ahead. 

2/ Could you please update bash file to be able to work for testnets as well?

hashedMae committed:
Merge branch 'feat/406-protocol-deploy-script' of https://github.com/ubiquity/ubiquity-dollar into feat/406-protocol-deploy-script

hashedMae committed:
fix: rebase
small corrections after rebase, remove tests of mock tokens

hashedMae committed:
fix(UbiquityChef): slither
old slither were somehow deleted in rebase

hashedMae committed:
fix(slither): constructors

hashedMae committed:
Update packages/contracts/src/dollar/core/CreditNFTManager.sol

Co-authored-by: アレクサンダー.eth <4975670+pavlovcik@users.noreply.github.com>

hashedMae committed:
fix: variable name

hashedMae committed:
fix(slither): immutable values

hashedMae committed:
Merge branch 'feat/406-protocol-deploy-script' of https://github.com/ubiquity/ubiquity-dollar into feat/406-protocol-deploy-script

hashedMae committed:
fix: PR review changes

hashedMae committed:
fix: bash and solidity script bugs

hashedMae committed:
feat: yarn workspace commands

hashedMae committed:
Update packages/contracts/src/dollar/core/TWAPOracleDollar3pool.sol

Co-authored-by: アレクサンダー.eth <4975670+pavlovcik@users.noreply.github.com>

hashedMae committed:
feat: PR comments

hashedMae committed:
Merge branch 'feat/406-protocol-deploy-script' of https://github.com/ubiquity/ubiquity-dollar into feat/406-protocol-deploy-script

hashedMae committed:
fix: natSpec comment typo

0x4007 commented on packages/contracts/package.json:
```diff
@@ -31,7 +31,9 @@
     "clean": "forge clean",
-    "purge": "yarn clean"
+    "purge": "yarn clean",
+    "anvil": "ts-node ./scripts/anvil/anvil.ts",
```
I think this is out of date from development because I updated `ts-node` to use `tsx` cause its faster.

0x4007 replied:
```suggestion
    "anvil": "tsx ./scripts/anvil/anvil.ts",
```

0x4007 commented on packages/contracts/package.json:
```diff
@@ -31,7 +31,9 @@
-    "purge": "yarn clean"
+    "purge": "yarn clean",
+    "anvil": "ts-node ./scripts/anvil/anvil.ts",
+    "run deploy:development": "bash ./scripts/bash/2_DevelopmentDeploy.sh"
```
Is this an issue to use the `bash` binary on windows?

hashedMae replied:
Might be. But who uses windows for dev work?

hashedMae replied:
Doing up a TS file instead cause I forgot about Windows users

hashedMae committed:
feat: remove bash script from package.json
Migrated bash commands to TS file

zgorizzo69 commented on .gitmodules:
```diff
@@ -13,6 +13,9 @@
 [submodule "packages/contracts/lib/forge-std"]
 	path = packages/contracts/lib/forge-std
 	url = https://github.com/foundry-rs/forge-std
+[submodule "packages/contracts/lib/abdk-libraries-solidity"]
```
nice

zgorizzo69 commented on packages/contracts/scripts/bash/2_DevelopmentDeploy.sh:
```diff
@@ -0,0 +1,15 @@
+
+
+# Impersonates an account with high 3crv token balance
+cast rpc anvil_impersonateAccount 0x4486083589A063ddEF47EE2E4467B5236C508fDe -r http://localhost:8545
```
as it can change shouldn't it be better to have the 3crv as an ENV ?

hashedMae replied:
Yes, that'd make more sense. The address in here is a 3CRV Whale. I already had to change it once while this PR was open as well.

hashedMae replied:
This shell script is also a vestige from before I created the TS script and can be deleted.

zgorizzo69 commented on package.json:
```diff
@@ -36,7 +36,7 @@
   },
   "lint-staged": {
-    "*.{js,jsx,ts,tsx,json,sol}": [
+    "*.{js,jsx,ts,tsx,json, sol}": [
```
?

0x4007 replied:
```suggestion
    "*.{js,jsx,ts,tsx,json,sol}": [
```

zgorizzo69 commented on packages/contracts/scripts/deploy/dollar/DevelopmentDeploy.ts:
```diff
@@ -0,0 +1,40 @@
+import { spawn } from "child_process";
+
+(async () => {
+  const command = spawn("cast", ["rpc", "anvil_impersonateAccount", "0x4486083589A063ddEF47EE2E4467B5236C508fDe", "-r", "http://localhost:8545"]);
```
same as `[2_DevelopmentDeploy.sh]` address should come from ENV

zgorizzo69 commented on packages/contracts/src/dollar/utils/SafeAddArray.sol:
```diff
@@ -1,38 +1,14 @@
-
     function add(uint256[] storage array, uint256 value) internal {
-        for (uint256 i; i < array.length; i++) {
+        for (uint256 i = 0; i < array.length; i++) {
```
Remove the redundant zero initialization uint256 i; instead of uint256 i = 0;   saves gas

0x4007 replied:
```suggestion
        for (uint256 i; i < array.length; i++) {
```

zgorizzo69 commented on packages/contracts/src/manager/facets/DiamondLoupeFacet.sol:
```diff
@@ -24,11 +24,14 @@ contract DiamondLoupeFacet is IDiamondLoupe, IERC165 {
             facets_[i].functionSelectors =
                 ds.facetFunctionSelectors[facetAddress_].functionSelectors;
+                unchecked {
+                    i++;
```
When the value of the post-loop increment/decrement is not stored or used in any calculations, the prefix increment/decrement operators (++i/--i) cost less gas PER LOOP than the postfix increment/decrement operators (i++/i--)

zgorizzo69 reviewed (commented):
running yarn update the yarn lock you should commit that updated yarn lock 
can you add a readme section to explain how the deployment works and show some examples

0x4007 commented on packages/contracts/src/dollar/core/UbiquityDollarManager.sol:
```diff
@@ -81,7 +75,7 @@ contract UbiquityDollarManager is AccessControl {
         // to be removed
 
         TWAPOracleDollar3pool oracle = TWAPOracleDollar3pool(twapOracleAddress);
-        oracle.update();
```
@zgorizzo69 any thoughts on this

zgorizzo69 replied:
@hashedMae I think we can keep it as a good practice just to make sure that the newly added oracle is updated 
@pavlovcik not that TWAP oracle is a facet in the diamond final implementation so updating its state will be done inside the diamond init

hashedMae committed:
Update package.json

Co-authored-by: アレクサンダー.eth <4975670+pavlovcik@users.noreply.github.com>

hashedMae committed:
Update packages/contracts/src/dollar/utils/SafeAddArray.sol

Co-authored-by: アレクサンダー.eth <4975670+pavlovcik@users.noreply.github.com>

hashedMae committed:
feat(deploy script): PR reccomendations

hashedMae committed:
Merge branch 'feat/406-protocol-deploy-script' of https://github.com/ubiquity/ubiquity-dollar into feat/406-protocol-deploy-script

hashedMae committed:
feat(deployment script): change readme and anvil.ts
-updated readme to explain local deployment with Solidity Scripting
-changed anvil start script to read RPC_URL value from .env

hashedMae committed:
fix(slither): error in SafeAddArrray.sol

hashedMae committed:
fix(slither): error in SafeAddArrray.sol

0x4007 commented on packages/contracts/package.json:
```diff
@@ -16,22 +16,24 @@
+    "test": "tsx ./scripts/runner/runner.ts",
+    "slither:run": "slither . --compile-force-framework foundry --config-file slither.config.json",
     "echidna": "echidna-test . --config echidna.config.yml",
     "prettier": "forge fmt",
```
I just realized that this could be causing the discrepancy with the formatting. Because this task is named "prettier" it is possible that lint-staged is calling this instead of what we expect.
```suggestion
```

zgorizzo69 replied:
if we want to use prettier then we need to update that otherwise it will format based on the settings in the foundry.toml and not inside the .prettierrc

0x4007 replied:
I just created a new issue because it looks like @hashedMae wants to leave this as is in this PR.

zgorizzo69 commented on packages/contracts/.env.example:
```diff
@@ -4,5 +4,14 @@ RPC_URL=
+ETHERSCAN_API_KEY=
+
+# Mnemonic for deployer account
+MNEMONIC =
```
by default can be the sahred mnemonic as stated inside the readme

hashedMae replied:
done and update `PRIVATE_KEY` and `ADMIN_ADDRESS` as well.

zgorizzo69 commented on packages/contracts/.env.example:
```diff
@@ -4,5 +4,14 @@ RPC_URL=
 PRIVATE_KEY=
 
+# Public key for deployer account
+PUBLIC_KEY= 
```
where is it used ? You should derive it directly from the private key

hashedMae replied:
It's used in the Solidity script. I don't know of anyway to derive an address from a private key within Solidity without creating a library to do the math for it. I should also rename this to something like `PUBLIC_ADDRESS` or `ADMIN_ADDRESS`

zgorizzo69 commented on README.md:
````diff
@@ -114,14 +114,19 @@ Make sure you are using the following network configuration:
-UAD_MANAGER_ADDRESS= # enter address here
-yarn hardhat --network localhost faucet --receiver $YOUR_ETH_ADDRESS --manager $UAD_MANAGER_ADDRESS
-```
+``` yarn workspace @ubiquity/contracts anvil ```
````
![image](https://user-images.githubusercontent.com/16853695/214290667-f9f284af-3ca7-4595-bc21-15404d5cf9f6.png)

hashedMae replied:
What version is your Forge? Sometimes need to upgrade `forgeup` to update.

hashedMae replied:
I had this working yesterday but is now throwing same error for me. Working on it.

zgorizzo69 commented on packages/contracts/scripts/deploy/dollar/DevelopmentDeploy.ts:
```diff
@@ -1,40 +1,50 @@
+    "cast",
+    [
+      "send",
+      "0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490",
```
this can also come from env if we want to deploy on other chain

hashedMae replied:
Yeah, I don't know why I didn't do that to start with. Definite oversight moment. This is also doing account impersonation so would only work on a local testnet.

hashedMae replied:
fixed

zgorizzo69 commented on packages/contracts/scripts/deploy/dollar/DevelopmentDeploy.ts:
```diff
@@ -1,40 +1,50 @@
+    {
+      stdio: "inherit",
+    }
+  );
```
it seems to me that all the commands should run one after another but as we call each one of the async functions without waiting we may have concurrency problems. I would recommend using this pattern instead

```
const command1 = async () => {...}
const command2 = async () => {...}
const command3 = async () => {...}
const main = async () => {
  await command1();
  await command2();
  await command3();
};

main();

```

hashedMae replied:
Thank you, my TS/JS is weak.  This is a definite improvement.

zgorizzo69 commented on packages/contracts/scripts/deploy/dollar/DevelopmentDeploy.ts:
```diff
@@ -1,40 +1,50 @@
-  command.stdout.on("data", (output: any) => {
-    console.log(output.toString());
-  });
+  const command = spawn(
```
maybe add some logging like  console.log("Sending 1000 USDC to the deployer"); for each steps

hashedMae replied:
Done

zgorizzo69 commented on packages/contracts/scripts/deploy/dollar/solidityScripting/01_UbiquityDollarManager.s.sol:
```diff
@@ -0,0 +1,20 @@
+    UbiquityDollarManager manager; 
+    function run() public virtual {
+
+        uint256 deployerPrivateKey = vm.envUint("PRIVATE_KEY");
```
they all inherits from constants where deployerPrivateKey is defined so we can remove it from all the run functions

hashedMae replied:
Thank you, I'd created a template and completely missed the redundancy.

hashedMae committed:
fix(solidity scripts): review comments
-code redundancy
-clean up whitespace

hashedMae committed:
fix(solidity scripting): pr comments

-change DevelopmentDeploy.ts to await each command
-rename 'PUBLIC_KEY' in .env to 'ADMIN_ADDRESS'

hashedMae committed:
fix(solidity scripting):  pr comments
-Changed DevelopmentDeploy.ts to use  from
-Added USD3CRV_TOKEN as .env value
-array cleanup in soldity scripts
-added prettier-plugin-solidity to yarn lock

hashedMae committed:
Merge https://github.com/ubiquity/ubiquity-dollar into feat/406-protocol-deploy-script

hashedMae committed:
merge development into 406

hashedMae committed:
Merge branch 'feat/406-protocol-deploy-script' of https://github.com/ubiquity/ubiquity-dollar into feat/406-protocol-deploy-script

hashedMae committed:
fix: contract dependency conflict

hashedMae committed:
fix: contract dependency conflict

hashedMae committed:
fix:broken tests

hashedMae committed:
fix: failing tests

zgorizzo69 reviewed (commented):
IMO we should add more explanation about how to deploy locally
especially the anvil script the output is misleading `using https://rpc.flashbots.net for anvil...` it should be something like forking from RPC_URL also we already have a bunch of url for forking mainnet inside scripts/runner/conf.ts why don't we use them if no RPC-URL or if the URL is not responding (cos yeah anvil script just stop with no error output if rpc is invalid..) ?

hashedMae committed:
feat(solidity scripting): anvil use rpc list
-Anvil uses the RPC list in conf.ts if RPC_URL in .env is not added

hashedMae committed:
fix(slither): unchecked transfer
-ignored return values from IERC20 transfer
-corrected dollarPriceReset and crvPriceReset in Staking

hashedMae committed:
fix(slither): multiplication on result of division
-DollarMintCalculator getDollarsToMint

hashedMae committed:
fix(slither): multiplication on result of division
-UbiquityChef pendingGovernance and _updatePool

hashedMae committed:
fix: redundant code
-UbiquityChef _updatePool

hashedMae committed:
fix: staking
-contract size exceeds max
